nexus/src/db/queries/network_interface.rs

```diff
@@ -979,16 +989,342 @@ impl QueryFragment<Pg> for InsertNetworkInterfaceQueryValues {
         out.push_identifier(dsl::ip::NAME)?;
         out.push_sql(", ");
         out.push_identifier(dsl::slot::NAME)?;
+        out.push_sql(", ");
+        out.push_identifier(dsl::is_primary::NAME)?;
         out.push_sql(") ");
         self.0.walk_ast(out)
     }
 }
 
+/// A small helper subquery that automatically assigns the `is_primary` column
+/// for a new network interface.
+///
+/// An instance with any network interfaces must have exactly one primary. (An
+/// instance may have zero interfaces, however.) This subquery is used to insert
+/// the value `true` if there are no extant interfaces on an instance, or
+/// `false` if there are.
+#[derive(Debug, Clone, Copy)]
+struct IsPrimaryNic {
+    instance_id: Uuid,
+}
+
+impl QueryId for IsPrimaryNic {
+    type QueryId = ();
+    const HAS_STATIC_QUERY_ID: bool = false;
+}
+
+impl QueryFragment<Pg> for IsPrimaryNic {
+    fn walk_ast<'a>(
+        &'a self,
+        mut out: AstPass<'_, 'a, Pg>,
+    ) -> diesel::QueryResult<()> {
+        out.push_sql("SELECT NOT EXISTS(SELECT 1 FROM");
+        NETWORK_INTERFACE_FROM_CLAUSE.walk_ast(out.reborrow())?;
+        out.push_sql(" WHERE ");
+        out.push_identifier(dsl::instance_id::NAME)?;
+        out.push_sql(" = ");
+        out.push_bind_param::<sql_types::Uuid, Uuid>(&self.instance_id)?;
+        out.push_sql(" AND ");
+        out.push_identifier(dsl::time_deleted::NAME)?;
+        out.push_sql(" IS NULL LIMIT 1)");
+        Ok(())
+    }
+}
+
+type InstanceFromClause = FromClause<db::schema::instance::table>;
+const INSTANCE_FROM_CLAUSE: InstanceFromClause = InstanceFromClause::new();
+
+/// Delete a network interface from an instance.
+///
+/// There are a few preconditions that need to be checked when deleting a NIC.
+/// First, the instance must currently be stopped, though we may relax this in
+/// the future. Second, while an instance may have zero or more interfaces, if
+/// it has one or more, exactly one of those must be the primary interface. That
+/// means we can only delete the primary interface if there are no secondary
+/// interfaces. The full query is:
+///
+/// ```sql
+/// WITH
+///     interface AS MATERIALIZED (
+///         SELECT CAST(IF(
+///             (
+///                 SELECT
+///                     NOT is_primary
+///                 FROM
+///                     network_interface
+///                 WHERE
+///                     id = <interface_id> AND
+///                     time_deleted IS NULL
+///             )
+///                 OR
+///             (
+///                 SELECT
+///                     COUNT(1)
+///                 FROM
+///                     network_interface
+///                 WHERE
+///                     instance_id = <instance_id> AND
+///                     name = <name>
```

Comment:
Why are we WHERE-ing on both ID and name?

Is the ID not sufficient?

Reply:
The `network_interface` table has a unique index on the tuple `(instance_id, name)`, with an additional `WHERE time_deleted IS NULL` condition. This is all spelled out here so that we can use that index, since I've generally found that CRDB won't use indexes like that without specifying the exact columns it is over.

Reply:
Ah, sorry I misread what this was referring to. That comment is actually wrong, we don't want to filter on the basis of the name of the interface. What we're looking for in this subquery is the existence of other interfaces on this instance. We allow deleting a target interface if (1) it is not the primary or (2) the target is the only interface. That latter check must be done with the instance ID, but neither needs nor currently uses the name. I'll check again that this hits the right index. Thanks for flagging it.